### DISCUSSION TITLE
Cross-platform engine build guidelines?

### DISCUSSION THREAD

**Discussion by afritz1** [jrouwe/JoltPhysics#1213] [Q&A]
I'm learning how to include this in my 3D engine and settled on making it a git submodule, but I want everyone who builds my engine to not get caught up in the specifics of building Jolt. Ideally I run my engine's CMakeLists.txt and Jolt gets compiled automatically.

Maybe my CMakeLists.txt should auto-select the script in `JoltPhysics/Build/`? Or should people manually build Jolt themselves using one of those, and rely on my CMakeLists.txt linking Jolt via `find_library()` or `add_library("Jolt.lib" STATIC)`? I like the idea of Jolt sharing the same Visual Studio solution as my code if it's going to be a static lib.

I also want all the special vector instructions and interprocedural optimization off for some build configurations but if I change that in my local Jolt repo then git says the submodule has uncommitted changes, which makes me think I need my own CMakeLists.txt for Jolt.

Thanks for any help, I'm not the build engineer type and am new to physics engines.

**Comment by jrouwe** [comment]
Is there any particular reason why you went for a sub module? When you check out a project with submodules you have to do an extra step in order to get them, which is why I think there's a lot of negativity around git sub modules.

Have you taken a look at https://github.com/jrouwe/JoltPhysicsHelloWorld? This is by far the easiest way of including a 3rd party module in your own cmake project and it is very hassle free for your end users as it automatically downloads Jolt in the background.

**Reply by afritz1** [reply]
Thanks for responding. I wasn't familiar with FetchContent which seems to be the better option. I did run the HelloWorld example previously but didn't realize it was a CMake demonstration too.

Everything builds now but I see a dozen `const std::system_error::vftable` linker warnings about those exception types being two different byte sizes. My engine has another static lib in addition to Jolt and it sounds like me setting C++20 globally isn't matching what Jolt uses.

**Reply by jrouwe** [reply]
Can you please paste the exact output that the compiler produces? I've not seen this before and I can mix c++17 and 20 without any issues in both clang and MSVC.

**Reply by afritz1** [reply]
I'll tinker with this more tonight since there's still more to figure out (compiling only works with my LTO build config), but here's the end of the output:
```
...
4>Main.cpp
4>LINK : warning LNK4044: unrecognized option '/MTd'; ignored
4>LINK : warning C4743: 'const std::system_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\file.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::_System_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\file.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::runtime_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\file.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\assets\citydatafile.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::system_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\directory.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::_System_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\directory.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::runtime_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\directory.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\assets\citydatafile.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::filesystem::filesystem_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\directory.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::system_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\path.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::_System_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\path.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\game\game.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::runtime_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\utilities\path.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\assets\citydatafile.cpp': 24 and 32 bytes
4>LINK : warning C4743: 'const std::runtime_error::`vftable'' has different size in 'c:\users\aaron\documents\github\opentesarena\components\archives\bsaarchive.cpp' and 'c:\users\aaron\documents\github\opentesarena\opentesarena\src\assets\citydatafile.cpp': 24 and 32 bytes
4>Generating code
4>Finished generating code
4>otesa.vcxproj -> C:\Users\Aaron\Documents\GitHub\OpenTESArena\build\ReleaseGeneric\otesa.exe
4>Done building project "otesa.vcxproj".
```

The two static libs are compiled before all that. I wonder if my global CMake build configs are causing a problem, since I know Jolt only has Debug/Release/Distribution, but it shows up in ReleaseGeneric.
```
2>components.vcxproj -> C:\Users\Aaron\Documents\GitHub\OpenTESArena\build\components\ReleaseGeneric\components.lib
3>Jolt.vcxproj -> C:\Users\Aaron\Documents\GitHub\OpenTESArena\build\_deps\joltphysics-build\ReleaseGeneric\Jolt.lib
```

I fixed the /MTd warning by putting `set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")` near the top of my CMakeLists.txt and that makes my non-LTO builds compile. Also the byte count warnings only seem to happen in my LTO build configs.

Second edit: I got it to compile in WSL Debian and on Raspberry Pi 4 which is a good sign.

**Reply by afritz1** [reply]
Things are going well, but I started getting a linker error on Linux when I inherited DebugRendererSimple in my [Renderer](https://github.com/afritz1/OpenTESArena/blob/00ba26326e063599ca06c4d395dd9ed39552f40e/OpenTESArena/src/Rendering/Renderer.h#L38). It's fine on Windows.
```
...
[ 99%] Building CXX object OpenTESArena/CMakeFiles/otesa.dir/src/Main.cpp.o
[100%] Linking CXX executable ../otesa
lto-wrapper: warning: using serial compilation of 105 LTRANS jobs
lto-wrapper: note: see the ‘-flto’ option documentation for more information
/usr/bin/ld: /tmp/ccHxNmE1.ltrans12.ltrans.o:(.data.rel.ro+0x78): undefined reference to `typeinfo for JPH::DebugRendererSimple'
collect2: error: ld returned 1 exit status
make[2]: *** [OpenTESArena/CMakeFiles/otesa.dir/build.make:4903: otesa] Error 1
make[1]: *** [CMakeFiles/Makefile2:187: OpenTESArena/CMakeFiles/otesa.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
Do I need more CMAKE_CXX_FLAGS/CMAKE_EXE_LINKER_FLAGS? This is the Jolt part of my top-level [CMakeLists.txt](https://github.com/afritz1/OpenTESArena/blob/00ba26326e063599ca06c4d395dd9ed39552f40e/CMakeLists.txt#L71):
```CMake
...
ADD_COMPILE_DEFINITIONS(JPH_NO_FORCE_INLINE) # Fix __forceinline compile error in LTO builds.
ADD_COMPILE_DEFINITIONS(JPH_DEBUG_RENDERER) # Enable Jolt Physics debug rendering in all builds.

# Include Jolt
FetchContent_Declare(
    JoltPhysics
    GIT_REPOSITORY "https://github.com/jrouwe/JoltPhysics"
    GIT_TAG cede24d2733a4a473c6d486650ca9b6d0481681a # 5.1.0
    SOURCE_SUBDIR "Build")
FetchContent_MakeAvailable(JoltPhysics)
```
and this is in the engine [CMakeLists.txt](https://github.com/afritz1/OpenTESArena/blob/00ba26326e063599ca06c4d395dd9ed39552f40e/OpenTESArena/CMakeLists.txt#L777):
```CMake
TARGET_INCLUDE_DIRECTORIES(otesa PUBLIC "${JoltPhysics_SOURCE_DIR}/..")
TARGET_LINK_LIBRARIES(otesa Jolt components ${EXTERNAL_LIBS})
SET_TARGET_PROPERTIES(otesa PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${OpenTESArena_BINARY_DIR})
```
Thanks.

**Reply by jrouwe** [reply]
Jolt by default compiles without C++ RTTI, your project is compiling with RTTI. Either turn it off for your own project or turn it on for Jolt using the `CPP_RTTI_ENABLED` cmake option.

**Reply by afritz1** [reply]
Ah thanks, completely missed that since I never think about using RTTI.
